Add support for comments and metadata

kenhkan opened the pull request:
This adds support for comments and metadata.

# Comments

Comments are parsed by looking for a 7-sided "line", as a speech bubble is 7-sided as produced by Draw.IO. The parsed outcome is a `(speech-bubble ...)`. See the file `hs-vsh/samples/comment.lisp`.

# Metadata

I updated the format of the metadata to be a full JSON that is compacted (i.e. no space or newline) for easy parsing, in the form of:

```
[
    {
        "dir": ...,
        "file": ...,
        "name": ...,
        "ref": ...,
        "repo": ...,
    },
    ...
]
```

Example used in the sample file:

```
[
    {
        "dir": "somewhere/",
        "file": "leaf_a.js",
        "name": "A",
        "ref": "master",
        "repo": "https://github.com/bmfbp/bmfbp.git"
    },
    {
        "dir": "somewhere/",
        "file": "leaf_b.js",
        "name": "B",
        "ref": "master",
        "repo": "https://github.com/bmfbp/bmfbp.git"
    }
]
```

Metadata are parsed by looking for a JSON string that matches with the form above.

The output is a `(metadata <x>)` where `<x>` is a double-quote-enclosed string that is a JSON.

See `hs-vsh/samples/metadata.lisp` for an example.

kenhkan commented:
@guitarvydas One question for you: Right now the parse outputs the metadata as a JSON string. Does it work better if it is parsed to facts? If so, what do you propose should be the format?

guitarvydas commented:
1. Can you change it so that hs-vsh-drawio-to-fb is turned into hs_vsh_drawio_to_fb?  (dashes become underscores).  Prolog chokes on dashes.  This matters if the file appears as the name of a kind. (Recent discovery)

2. As long as the comments and metadata become Lisp forms, that is enough for now.  The toprolog.lisp file can mangle the output when we decide how we want to build the ide components.

guitarvydas commented:
The .svg files in samples will not load in my draw.io.

kenhkan commented:
@guitarvydas 

> Can you change it so that hs-vsh-drawio-to-fb is turned into hs_vsh_drawio_to_fb? (dashes become underscores). Prolog chokes on dashes. This matters if the file appears as the name of a kind. (Recent discovery)

Added https://github.com/bmfbp/bmfbp/pull/69/commits/b2371cd58541f99148cd03d3a179e715465eac43. Run `stack install` again to build the executable.

> As long as the comments and metadata become Lisp forms, that is enough for now. The toprolog.lisp file can mangle the output when we decide how we want to build the ide components.

What is Lisp forms in your mind?

For example, right now metadata is output as something like:

```
((rect 20.0 0.0 80.0 80.0) (translate (55.5 34.5) ((translate (5.0 12.0) ("A")))) (rect 240.0 0.0 80.0 80.0) (translate (275.5 34.5) ((translate (5.0 12.0) ("B")))) (line (absm 100.0 40.0) (absl 233.63 40.0)) (line (absm 238.88 40.0) (absl 231.88 43.5) (absl 233.63 40.0) (absl 231.88 36.5) (Z)) (rect 0.0 170.0 460.0 170.0) (translate (1.5 242.5) ((translate (229.0 19.0) ((metadata "[{\"dir\":\"somewhere/\",\"name\":\"A\",\"repo\":\"https://github.com/bmfbp/bmfbp.git\",\"file\":\"leaf_a.js\",\"ref\":\"master\"},{\"dir\":\"somewhere/\",\"name\":\"B\",\"repo\":\"https://github.com/bmfbp/bmfbp.git\",\"file\":\"leaf_b.js\",\"ref\":\"master\"}]"))))))
```

Is this sufficient?

> The .svg files in samples will not load in my draw.io.

Do you need to load it in draw.io? Or do you just need to view it?

I just realized that draw.io doesn't read the SVG that it outputs unless an option is selected, which renders the output a non-standard SVG with "compressed" content.

guitarvydas commented:
I would prefer to load it into Draw.IO.  I might want to change it / play with it to figure out how to compile it.

Lisp (metadata "<some long string>") is good enough, for now.   The example looks OK.

kenhkan commented:
@guitarvydas I added the source files (*.drawio). Open the diagrams using the .drawio files and parse using the .svg files.

kenhkan commented:
@guitarvydas 

I renamed `name` to `kindName` and added `kindType` in the metadata of the test cases as we discussed today.

I also updated `all_diagrams.svg` with `all_diagrams.drawio` so that you can open the diagram in Draw.IO.

guitarvydas commented:
The dash in "speech-bubble" needs to be removed --> "speechbubble".  I thought that I removed it from src/Parser/DrawIO.hs, but I keep seeing speech-bubble (with the dash).

guitarvydas commented:
I see that there is some Lisp after speech-bubble gets generated.  I can fix this in lisp, so it's not blocking me.  At some point I need a lesson on how to change DrawIO.hs and have it rebuild.

guitarvydas commented:
update: The Makefile in hs-vsh cp's the -'ed version instead of the _'ed version of the parser.

guitarvydas commented:
update: hs_vsh_drawio_to_fb <svgc.svg appears to produce a speech bubble where there is none (svgc.svg is found in branch js-test2 bmfbp/svg/flat-compiler/svgc.svg).

guitarvydas commented:
update: I see Q commands in the path(s) on that diagram - they are unsupported in DrawIO.hs.  I'll see if I can clean up the diagram...

guitarvydas commented:
update: Svgc.svg had "rounded" arrows.  I made them all "square" and got further.  UnsupportedPathCommand should probably result in something drastic that makes hs_... fail loudly.

kenhkan commented:
> The dash in "speech-bubble" needs to be removed --> "speechbubble". I thought that I removed it from src/Parser/DrawIO.hs, but I keep seeing speech-bubble (with the dash).
>
> I see that there is some Lisp after speech-bubble gets generated. I can fix this in lisp, so it's not blocking me. At some point I need a lesson on how to change DrawIO.hs and have it rebuild.

Updated. Please test again.

> update: The Makefile in hs-vsh cp's the -'ed version instead of the _'ed version of the parser.

This is updated as well. Please verify.

> update: hs_vsh_drawio_to_fb <svgc.svg appears to produce a speech bubble where there is none (svgc.svg is found in branch js-test2 bmfbp/svg/flat-compiler/svgc.svg).
> 
> update: I see Q commands in the path(s) on that diagram - they are unsupported in DrawIO.hs. I'll see if I can clean up the diagram...
>
> update: Svgc.svg had "rounded" arrows. I made them all "square" and got further. UnsupportedPathCommand should probably result in something drastic that makes hs_... fail loudly.

It doesn't look like it's a problem now. Is that so?

> (From Slack)
>
> hmmm, my version of hs_vsh_drawio_to_fb + draw.io + hs-vsh/samples/metadata.svg does not produce the same results as you show (instead of “metadata”, I have “<span>…“).  Maybe we are using different versions of draw.io, maybe I’m doing something unanticipated, maybe I broke hs_vsh_drawio_to_fb while playing with it yesterday (I’m on branch js-tests2 and that contains my lates hs_vsh_drawio_to_fb).
> 
> Hmmm, I just re-pulled the branch add-support-for-comments-and-metadata, ran “stack install” and “make”, then ran “~/.local/bin/hs_vsh_drawio_to_fb <metadata.svg >pt4” and got the result with <span> instead of metadata.  My version of draw.io is 10.7.3.  It looked like the hs_vsh_drawio_to_fb was timestamped correctly.

This should be fixed. Not sure why I didn't get `<span>` before. Can you test?

guitarvydas commented:
Yes, I'm seeing the above, now.  Thanks.
